Cpp_hands_on_programming/lib/abstract/Ipc_method.cc

```diff
@@ -0,0 +1,60 @@
+#include "Ipc_method.h"
+#include <iostream>
+
+int Ipc_method::getFileSize(std::string filename) {
+    int size;
+    std::ifstream in(filename, std::ios::binary);
+	if (!in) {
+		perror("getFileSize");
+		exit(EXIT_FAILURE);
+	}
+	in.seekg(0, std::ios::end);
+	size = in.tellg();
+    in.close();
+    return size;
+}
+
+int Ipc_method::readFile(std::string filename, int offset, std::vector<char> &buffer, int size) {
+    std::ifstream in(filename, std::ios::binary);
+	if (!in.is_open()) {
+		perror("ifstream::open\n");
+		return -1;
+	}
+	// move file position according to offset
+	in.seekg(offset, std::ios::beg);
+	if (!in) {
+		perror("seekg\n");
+		return -1;
+	}
+
+	in.read(&buffer[0], size);
+	// check read status, in.gcount condition set to make last read possible
+	if (!in && in.gcount() == size) {
```

Comment:
why do you think, that if there will be actual failure during read, the gcount() will equal size?

Reply:
My idea is to allow the last chunk of data (if its size less than the default size) being processed without throwing error.

 As quoted in cplusplus.com "If the input sequence runs out of characters to extract (i.e., the end-of-file is reached) before n characters have been successfully read, the array pointed to by s contains all the characters read until that point, and both the [eofbit](https://www.cplusplus.com/ios_base::eofbit) and [failbit](https://www.cplusplus.com/ios_base::failbit) flags are set for the stream."